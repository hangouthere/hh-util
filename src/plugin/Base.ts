import { AlreadyEnabledError, NotEnabledError } from './Error';
import PluginHost from './Host';

export enum PluginType {
  SERVICE = 'SERVICE',
  GUI_COMMON = 'GUI_COMMON',
  GUI_MAIN = 'GUI_MAIN'
}

export enum PluginStatus {
  STOPPED,
  STARTED,
  PAUSED,
  ERRORED
}

type Behavior = string;
type AccessorAlias = string;

export interface PluginManifest {
  // Accessor Manifest Config
  accessors?: Record<AccessorAlias, Behavior>;
  behaviours?: Array<Behavior>;
  pluginPath: string;
  name: string;
  type: PluginType;
  version: string;
  // views?: object; //TODO: Type this out with sub-views
}

export type AccessorAliasMap = Record<AccessorAlias, Array<PluginBase>>;

export interface IPluginOptions {
  host: PluginHost;
  manifest: PluginManifest;
  // Resolved Accessors, from DepMgr
  accessors: AccessorAliasMap;
}

export default class PluginBase {
  static ROOT_KEY: string = 'Plugins';

  accessors: AccessorAliasMap;
  host: PluginHost;
  manifest: PluginManifest;
  enabled: boolean = false;
  status: PluginStatus = PluginStatus.STOPPED;

  constructor(options: IPluginOptions) {
    this.accessors = options.accessors!;
    this.host = options.host;
    this.manifest = Object.freeze(options.manifest);
    this.enabled = this.getSetting('enabled') || false;

    if (this.manifest.behaviours && false === this.manifest.behaviours.includes(this.manifest.name)) {
      this.manifest.behaviours.push(this.manifest.name);
    }
  }

  async start(): Promise<any> {
    if (false === this.enabled) {
      this.status = PluginStatus.ERRORED;
      throw new NotEnabledError(this.manifest.name);
    }

    this.status = PluginStatus.STARTED;

    return Promise.resolve(this);
  }

  async stop(): Promise<any> {
    if (false === this.enabled) {
      this.status = PluginStatus.ERRORED;
      throw new NotEnabledError(this.manifest.name);
    }

    this.status = PluginStatus.STOPPED;

    return Promise.resolve(this);
  }

  async restart(): Promise<any> {
    return this.stop().then(() => this.start());
  }

  async enable(): Promise<any> {
    if (true === this.enabled) {
      this.status = PluginStatus.ERRORED;
      throw new AlreadyEnabledError(this.manifest.name);
    }

    this.enabled = true;

    return Promise.resolve(this);
  }

  async disable(): Promise<any> {
    if (false === this.enabled) {
      this.status = PluginStatus.ERRORED;
      throw new NotEnabledError(this.manifest.name);
    }

    await this.stop();

    this.enabled = false;

    return Promise.resolve(this);
  }

  getSetting(settingPath: string) {
    return this.host.getOption(`${PluginBase.ROOT_KEY}.${this.manifest.name}.${settingPath}`);
  }

  setSetting(settingPath: string, val: any) {
    this.host.setOption(`${PluginBase.ROOT_KEY}.${this.manifest.name}.${settingPath}`, val);
  }
}
