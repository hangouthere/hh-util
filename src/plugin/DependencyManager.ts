import uniq from 'lodash.uniq';
import { EventEmitter } from 'node:events';
import path from 'node:path';
import Logger from '../core/Logger.js';
import Base, { type AccessorAliasMap, type Manifest } from './Base.js';
import Host from './Host.js';
import { type ILoaderMeta } from './Loader.js';

const MODULE_NAME: string = path.basename(__filename, path.extname(__filename));
const log = Logger.extend(MODULE_NAME);

export type PluginNameMap = Record<string, ILoaderMeta>;

export type BehaviourNameMap = Record<string, Array<ILoaderMeta>>;

export type PluginLoadMap = Record<string, Promise<Base>>;

export type PluginFailMap = Record<string, DependencyChainError>;

export type PluginChainMap = Record<string, Array<string>>;

export class DependencyChainError extends Error {
  origError: Error;
  chainMap: Array<string> = [];

  constructor(pluginName: string, origError: Error) {
    super(`[DependencyChainError] - ${pluginName} has failed an entire dependency chain`);

    this.origError = origError;
    this.chainMap.push(pluginName);

    Error.captureStackTrace(this, DependencyChainError);
  }
}

export default class DependencyManager extends EventEmitter {
  behaviourNameMap: BehaviourNameMap = {};
  chainMap: PluginChainMap = {};
  failMap: PluginFailMap = {};
  loadMap: PluginLoadMap = {};
  pluginNameMap: PluginNameMap = {};

  static getFullyQualifiedName(resource: ILoaderMeta | Base) {
    return `${resource.manifest.name}-v${resource.manifest.version}`;
  }

  constructor(public pluginHost: Host) {
    super();

    let loaded = 0;
    let total = 0;
    this.on('detected', (names: Array<string>) => {
      total = names.length;
      log(`Detected ${total} plugins`);
    });

    this.on('loaded', (pluginInstance: Base) => {
      loaded++;
      log(`(${loaded}/${total}) - Loaded: ${pluginInstance.manifest.name}`);
    });
  }

  /**
   * Load Plugin Definitions into the Dependency tree for auto-injection
   */
  async loadPluginDefinitions(loaderInfos: Array<ILoaderMeta>): Promise<PluginLoadMap> {
    const pluginNameMap = this.buildPluginNameMap(loaderInfos);
    // Assume Plugin Names as Accessor Names for newly loaded Plugins
    const pluginNames = Object.keys(pluginNameMap);

    this.emit('detected', pluginNames);

    // Plugin names already loaded result in an error
    for (let name of pluginNames.concat()) {
      if (this.loadMap.hasOwnProperty(name)) {
        // Remove from known Plugin names and new name map to avoid further processing
        pluginNames.splice(pluginNames.indexOf(name), 1);
        delete pluginNameMap[name];
        this.failMap[name] = new DependencyChainError(name, new Error('Already Loaded'));
      } else {
        this.pluginNameMap[name] = pluginNameMap[name];
      }
    }

    // Blind overwrite of previously named mappings with new mapping.
    // This should normally only be due to newly loaded Plugins adding their behaviour to a previous list.
    const behaviourNameMap = this.buildBehaviourNameMap(loaderInfos);
    const behaviourNames = Object.keys(behaviourNameMap);
    behaviourNames.forEach(name => (this.behaviourNameMap[name] = behaviourNameMap[name]));

    const chainMap = this.buildAccessorChainMap(pluginNames);
    const chainNames = Object.keys(chainMap);
    chainNames.forEach(name => (this.chainMap[name] = chainMap[name]));

    await this.loadChainMap(pluginNames);

    return this.loadMap;
  }

  buildPluginNameMap(loaderMetas: Array<ILoaderMeta>): PluginNameMap {
    const hasDupe: Array<string> = [];

    const pluginMap = loaderMetas.reduce<PluginNameMap>((map, loaderMeta: ILoaderMeta) => {
      const name = DependencyManager.getFullyQualifiedName(loaderMeta);
      // Semaphore, must be handled outside of reduction or causes an uncaught exception
      if (map[name]) {
        hasDupe.push(name);
      } else {
        map[name] = loaderMeta;
      }

      return map;
    }, {});

    //@ts-ignore:next-line
    if (0 < hasDupe.length) {
      throw new Error(`${MODULE_NAME} - [DuplicateName] Duplicate Plugins detected: ${hasDupe.join(', ')}`);
    }

    return pluginMap;
  }

  buildBehaviourNameMap(pluginLoaderMetas: Array<ILoaderMeta>): BehaviourNameMap {
    let retMap: BehaviourNameMap = {};

    return pluginLoaderMetas.reduce<BehaviourNameMap>((map, pluginClass: ILoaderMeta) => {
      const { behaviours } = pluginClass.manifest as Manifest;
      const name = DependencyManager.getFullyQualifiedName(pluginClass);
      if (behaviours) {
        behaviours.forEach(behaviour => {
          map[behaviour] = map[behaviour] || [];
          (map[behaviour] as Array<Object>).push(pluginClass);
        });
      }
      // Ensure the Plugin name listed as a Behaviour
      map[name] = map[name] || [];
      (map[name] as Array<Object>).push(pluginClass);
      return map;
    }, retMap);
  }

  getAccessorNamesForPlugins(accessorGroup: Array<ILoaderMeta>): Array<string> {
    const names = accessorGroup.reduce((names: Array<string>, accessorPlugin) => {
      if (accessorPlugin.manifest.accessors) {
        const subAccessorNames = Object.values(accessorPlugin.manifest.accessors);
        names = subAccessorNames.concat(names);
      }

      return names;
    }, []);

    return uniq(names).sort();
  }

  accessorNameToPluginNames(accessorName: string): Array<string> {
    if (!this.behaviourNameMap[accessorName]) {
      throw new Error(`${MODULE_NAME} - [DependencyMissing] Missing Dependency for Accessor: ${accessorName}`);
    }

    return uniq(
      this.behaviourNameMap[accessorName].map(loaderMeta => DependencyManager.getFullyQualifiedName(loaderMeta))
    ).sort();
  }

  buildAccessorChainMap(
    accessorNames: Array<string>,
    chainMap: PluginChainMap = {},
    pluginChain: Array<string> = []
  ): PluginChainMap {
    for (let accessorName of accessorNames) {
      // Skip if this has been processed before
      if (chainMap[accessorName]) {
        continue;
      }

      // Look up accessor for all Plugins that satisfy it
      const behaviourGroup = this.behaviourNameMap[accessorName];
      // Break immediately if Accessor/Plugin is missing
      if (!behaviourGroup || 0 == behaviourGroup.length) {
        throw new Error(`${MODULE_NAME} - [PluginMissing] No reference found for Accessor: ${accessorName}`);
      }

      // Get all Plugin Names of the iterated Accessor
      const accessorPluginNames = new Set(this.accessorNameToPluginNames(accessorName));
      // Get all dependency Accessor Names of the Plugins in the behaviourGroup
      const dependencyAccessorNames = this.getAccessorNamesForPlugins(behaviourGroup);
      // Convert all Accessor Names into their associative Plugin Names
      const dependencyPluginNames = dependencyAccessorNames
        ? dependencyAccessorNames.reduce<Array<string>>(
            (plugins: Array<string>, name: string) => plugins.concat(this.accessorNameToPluginNames(name)),
            []
          )
        : [];
      // Break if self referenced
      const pluginsSelfRef = dependencyPluginNames.filter(pluginName => accessorPluginNames.has(pluginName));
      if (0 !== pluginsSelfRef.length) {
        throw new Error(`${MODULE_NAME} - [SelfReference] Invalid Self Reference: ${[...pluginsSelfRef].join(', ')}`);
      }

      // Break if cyclic reference
      const pluginsCyclical = dependencyPluginNames.filter(pluginName => pluginChain.includes(pluginName));
      if (0 !== pluginsCyclical.length) {
        throw new Error(`${MODULE_NAME} - [Cyclical] Dependency Detected: ${pluginChain.join(' -> ')}`);
      }

      // Sub-Accessors exist, so we need to recursively build their chain maps
      if (0 !== dependencyPluginNames.length) {
        // If existing, update the known Accessors for the chosen
        //  `associatedAccessorName` to be loaded after the newly found ones
        chainMap[accessorName] = uniq(dependencyPluginNames);

        // Recursively detect depencency tree
        //   - Make sure we know about ourselves in the Plugin chain, but don't update the
        //     original for subsequent sibling iterations!
        this.buildAccessorChainMap(dependencyPluginNames, chainMap, pluginChain.concat(accessorName));
      }
    }

    return chainMap;
  }

  async loadChainMap(pluginNames: Array<string>): Promise<Array<Base>> {
    const pluginLoads = pluginNames.map(async pluginName => {
      // Already loaded, let's shortcut
      if (this.loadMap.hasOwnProperty(pluginName)) {
        return this.loadMap[pluginName];
      }

      let loadChain: any = Promise.resolve();

      const pluginChainList = this.chainMap[pluginName];

      // If we have a dependency Plugin Chain, recursively process it...
      if (pluginChainList) {
        loadChain = loadChain.then(() => this.loadChainMap(pluginChainList));
      }

      loadChain = loadChain.then(() => this.loadPluginInstance(pluginName));

      this.loadMap[pluginName] = loadChain;

      return loadChain;
    });

    return await Promise.all(pluginLoads);
  }

  async mapAccessorAliases(accessorAliasDefinitions: { [accessorName: string]: string }): Promise<AccessorAliasMap> {
    const aliasNames = Object.keys(accessorAliasDefinitions);
    const map: AccessorAliasMap = {};

    for (let alias of aliasNames) {
      const pluginNames: Array<string> = this.accessorNameToPluginNames(accessorAliasDefinitions[alias]);

      map[alias] = await Promise.all(pluginNames.map(pluginName => this.loadMap[pluginName]));
    }

    return map;
  }

  async loadPluginInstance(pluginName: string): Promise<Base> {
    try {
      const loaderMeta = this.pluginNameMap[pluginName];
      const accessorAliasMap = loaderMeta.manifest.accessors
        ? await this.mapAccessorAliases(loaderMeta.manifest.accessors)
        : null;

      // Previously loaded Plugins re-use instances for subsequent requests
      // Not loaded, so we instantiate and load!
      const pluginInstance: Base = new (loaderMeta.pluginDefinition as any)({
        host: this.pluginHost,
        manifest: loaderMeta.manifest,
        accessors: accessorAliasMap
      });

      const promiseChain = Promise.resolve()
        .then(() => pluginInstance.enable())
        .then(() => pluginInstance.start())
        .then(() => this.emit('loaded', pluginInstance))
        .then(() => pluginInstance);

      return promiseChain;
    } catch (err) {
      const errCasted = err as Error;

      log('Error enabling/starting Plugin...', errCasted);

      const newError = new DependencyChainError(pluginName, errCasted);
      this.failMap[pluginName] = newError;
      throw newError;
    }
  }

  //TODO Finish unloading of plugins thoroughly
  async unloadPluginInstance(pluginName: string): Promise<Base> {
    try {
      const pluginInstance = await this.loadMap[pluginName];

      await pluginInstance.stop();

      // const accessorNames = this.getAccessorNamesForPlugins([pluginInstance]);

      delete this.loadMap[pluginName];
      delete this.pluginNameMap[pluginName];
      // accessorNames.forEach(n => this.behaviourNameMap[n].splice(this.behaviourNameMap[n].indexOf(n), 1));

      return pluginInstance;
    } catch (err) {
      log('Error stopping Plugin...', err);

      throw err;
    }
  }
}
