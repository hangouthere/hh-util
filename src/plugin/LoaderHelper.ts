class LoaderHelper {
  resolve(modulePath: string) {
    return require.resolve(modulePath);
  }

  get cache() {
    return require.cache;
  }
}

export default new LoaderHelper();
