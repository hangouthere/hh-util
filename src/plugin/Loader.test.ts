import * as fs from 'fs';
import { PluginType, type PluginManifest } from './Base';
import DependencyManager from './DependencyManager';
import PluginHost from './Host';
import PluginLoader from './Loader';
import LoaderHelper from './LoaderHelper';

expect.addSnapshotSerializer({
  test: val => typeof val === 'string',
  print: val => (val as string).replace(/\\/g, '/')
});

jest.mock('fs');

jest.mock('../core/Logger', () => {
  const Mock: any = jest.fn();
  // const Mock: any = console.log;
  // Mock.extend = msg => {
  //   const prevExtend = Mock.extend;
  //   const extended = Mock.bind(console, msg, '->');
  //   extended.extend = prevExtend;
  //   return extended;
  // };

  Mock.extend = jest.fn().mockReturnValue(Mock);

  return Mock;
});

jest.mock('./Host', () => {
  const Mock = {
    setOption: jest.fn()
  };

  return jest.fn(() => Mock);
});

jest.mock('./DependencyManager', () => {
  const Mock = {
    loadPluginDefinitions: jest.fn().mockReturnValue(Promise.resolve('depManagerLoaded'))
  };

  return jest.fn(() => Mock);
});

jest.mock('./LoaderHelper', () => {
  const Mock = {
    resolve: jest.fn(),
    cache: {}
  };

  return Mock;
});

jest.mock(
  'fakePlugin',
  () => {
    const MockInstance = {
      enable: () => {
        MockInstance.enabled = true;
        return Promise.resolve();
      },
      enabled: false
    };

    const Mock: any = {
      default: jest.fn(() => MockInstance)
    };

    return Mock;
  },
  { virtual: true }
);

jest.mock(
  'fakePluginInvalid',
  () => {
    const MockInstance = {
      enable: () => {
        MockInstance.enabled = true;
        return Promise.resolve();
      },
      enabled: false
    };

    const Mock: any = {
      default: jest.fn(() => MockInstance)
    };

    return Mock;
  },
  { virtual: true }
);

jest.mock(
  'fakePluginBroken',
  () => {
    const MockInstance = {
      enable: () => {
        MockInstance.enabled = true;
        return Promise.resolve();
      },
      enabled: false
    };

    const Mock: any = {
      default: jest.fn(() => MockInstance)
    };

    return Mock;
  },
  { virtual: true }
);

describe('PluginLoader', () => {
  let loader: PluginLoader;
  let host: PluginHost;
  let depMgr: DependencyManager;

  let manifest: PluginManifest;

  beforeEach(() => {
    loader = new PluginLoader({ pluginPaths: ['fakePluginPath'], pluginTypes: [PluginType.SERVICE] });
    host = new PluginHost();
    depMgr = new DependencyManager(host);

    manifest = {
      accessors: { plug2: 'plugin2-behaviour' },
      behaviours: ['plugin1-behaviour', 'common1'],
      name: 'Plugin1',
      pluginPath: '',
      version: '0.0.1',
      type: PluginType.SERVICE
    };

    jest.spyOn(console, 'log').mockImplementation(jest.fn());
  });

  describe('Plugin Loading', () => {
    it('should unload invalid plugins from memory', () => {
      jest
        .spyOn(LoaderHelper, 'resolve')
        .mockReturnValueOnce('fakePluginInvalid')
        .mockReturnValueOnce('fakePluginInvalid2');

      LoaderHelper.cache['fakePluginInvalid'] = { children: [{ filename: 'fakePluginInvalid', children: [] }] } as any;
      LoaderHelper.cache['fakePluginInvalid2'] = { children: [] } as any;

      loader.unloadModule('fakePluginInvalid', '');

      expect(LoaderHelper.cache).not.toHaveProperty('fakePluginInvalid');
    });

    it('should skip unloading Plugins outside of the namespace', () => {
      LoaderHelper.cache['fakePluginInvalid'] = { children: [] } as any;

      loader.unloadModule('fakePluginInvalid', 'basePath');

      expect(LoaderHelper.cache).toHaveProperty('fakePluginInvalid');
    });
  });

  describe('Plugin Scanning', () => {
    it('should return properly found Plugins in a Directory', async () => {
      jest
        .spyOn(fs, 'readdir')
        .mockImplementationOnce((path, _opts, cb) => {
          return cb(null, ['file1', 'file2', 'package.json'] as any);
        })
        .mockImplementationOnce((path, _opts, cb) => {
          return cb(null, ['file3', 'file4', 'package.json'] as any);
        });

      jest.spyOn(fs, 'readFile').mockImplementation((_path, cb) => {
        return cb(null, Buffer.from('{ "MANIFEST": { "type": "SERVICE" }, "name":"FakePlugin", "version": "1.0.0" }'));
      });

      // Use the impls above for these
      await expect(loader.findPluginManifest('tsPluginDir')).resolves.toMatchSnapshot();
      await expect(loader.findPluginManifest('jsPluginDir')).resolves.toMatchSnapshot();
      // Default response doesn't have a plugin file
      await expect(loader.findPluginManifest('noPluginPath')).resolves.toBeUndefined();
    });

    it("should return an undefined value when a Plugin isn't properly identified", async () => {
      jest.spyOn(fs, 'readdir').mockImplementationOnce((_path, _opts, cb) => {
        return cb(null, ['file1', 'file2', 'package.json'] as any);
      });

      jest.spyOn(fs, 'readFile').mockImplementation((_path, cb) => {
        return cb(null, Buffer.from('{  "name":"NotAPlugin", "version": "1.0.0" }'));
      });

      // Use the impls above for these
      await expect(loader.findPluginManifest('tsPluginDir')).resolves.toBeUndefined();
    });

    it('should skip plugins not meant for the current runtime (aka domain)', async () => {
      jest.spyOn(fs, 'readdir').mockImplementationOnce((_path, _opts, cb) => {
        return cb(null, ['file1', 'file2', 'package.json'] as any);
      });

      jest.spyOn(fs, 'readFile').mockImplementation((_path, cb) => {
        return cb(null, Buffer.from('{ "MANIFEST": { "type": "GUI" }, "name":"FakePlugin", "version": "1.0.0" }'));
      });

      // Use the impls above for these
      await expect(loader.findPluginManifest('tsPluginDir')).resolves.toBeUndefined();
    });

    it('should find all Plugins in a Scan Directory', async () => {
      jest
        .spyOn(fs, 'readdir')
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, ['pluginDir1', 'pluginDir2'] as any);
        })
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, ['index.plugin.ts'] as any);
        })
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, ['index.plugin.js'] as any);
        });

      jest.spyOn(fs, 'stat').mockImplementation((_path, _opts, cb) => {
        return cb(null, { isDirectory: jest.fn().mockReturnValue(true) } as any);
      });

      jest.spyOn(loader, 'findPluginManifest').mockResolvedValue(manifest);

      await expect(loader.detectPluginPackages('anyFakeDir')).resolves.toMatchSnapshot();
    });

    it('should skip non-directories in a Scan Directory', async () => {
      jest
        .spyOn(fs, 'readdir')
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, ['notAFolder', 'pluginDir1'] as any);
        })
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, ['index.plugin.ts'] as any);
        });

      jest
        .spyOn(fs, 'stat')
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, { isDirectory: jest.fn().mockReturnValue(false) } as any);
        })
        .mockImplementationOnce((_path, _opts, cb) => {
          return cb(null, { isDirectory: jest.fn().mockReturnValue(true) } as any);
        });

      await expect(loader.detectPluginPackages('anyFakeDir')).resolves.toMatchSnapshot();
    });

    it('should find all Plugins in all Scan Directories', async () => {
      jest.spyOn(loader, 'detectPluginPackages').mockResolvedValue([manifest]);

      await expect(loader.scanForPlugins(['fakePath'])).resolves.toMatchSnapshot();
    });

    it('should throw an error when failing to detect plugins', async () => {
      jest.spyOn(loader, 'detectPluginPackages').mockImplementation(() => {
        throw new Error('TestError: Failed to detect');
      });

      try {
        await loader.scanForPlugins(['fakePath']);
      } catch (err) {
        expect(err).toBeDefined();
        expect((err as Error).toString()).toMatch('Failed to detect');
      }
    });
  });

  describe('PluginLoader Initialization', () => {
    it('should load the plugin from the disk', async () => {
      await loader.init();

      expect(depMgr.loadPluginDefinitions).toHaveBeenCalled();
    });

    it('should utilize the Dependency Manager to load Plugins', async () => {
      jest.spyOn(loader, 'scanForPlugins').mockResolvedValue([manifest]);

      await loader.init();

      expect(depMgr.loadPluginDefinitions).toHaveBeenCalled();
    });

    it('should throw an error when initializing', async () => {
      jest.spyOn(loader, 'scanForPlugins').mockImplementationOnce(path => {
        throw new Error('Test Error: PluginLoad');
      });

      await expect(loader.init()).rejects.toThrow('Test Error: PluginLoad');
    });
  });
});
