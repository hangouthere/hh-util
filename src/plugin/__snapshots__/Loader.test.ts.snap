// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginLoader Plugin Scanning should find all Plugins in a Scan Directory 1`] = `
Array [
  Object {},
  Object {},
]
`;

exports[`PluginLoader Plugin Scanning should find all Plugins in all Scan Directories 1`] = `
Array [
  plugin1/index.plugin.ts,
  plugin2/index.plugin.ts,
]
`;

exports[`PluginLoader Plugin Scanning should return properly found Plugins in a Directory 1`] = `
Object {
  name: FakePlugin,
  pluginPath: tsPluginDir,
  type: SERVICE,
  version: 1.0.0,
}
`;

exports[`PluginLoader Plugin Scanning should return properly found Plugins in a Directory 2`] = `
Object {
  name: FakePlugin,
  pluginPath: jsPluginDir,
  type: SERVICE,
  version: 1.0.0,
}
`;

exports[`PluginLoader Plugin Scanning should skip non-directories in a Scan Directory 1`] = `Array []`;
